gcs_light_client.mediawiki

```diff
@@ -0,0 +1,992 @@
+    let F = N * P
+
+    let hashed_items = []
+    for raw_item in raw_items:
+        let hashed_item = (siphash_key(siphash_key, raw_item) * F) >> 64
+        hashed_items.append(hashed_item)
+
+    hashed_items.sort()
+
+    gcs_compress(hashed_items, fp)
+</pre>
+
+==== Filter Capability Querying ====
+
+As it's feasible that in the future, this document is extended to encompass
+additional filter encoding algorithms or filter contents, we define a new p2p
+message that allows light clients to ascertain which filters a node supports.
+
+The <code>getcftypes</code> message is an ''empty message'' whose command string is:
+<code>getcftypes</code>
+
+A full-node that receives a <code>getcftypes</code> message should respond with a
+<code>cftypes</code> message which is defined as follows:
+
+{| class="wikitable"
+! Field Size
+! Description
+! Data Type
+! Comments
+|-
+| Var-Int
+| NumFilters
+| uint64
+| The number of supported filters.
+|-
+| NumFilters
+| SupportedFilters
+| [NumFilterBytes]byte
+| A byte slice with each byte denoting a supported filter type
+|}
+
+
+==== Compact Filter Header Chain ====
+
+As the filters described in this BIP ''are not'' consensus critical, meaning
+each filter is validated by full-nodes and committed into blocks by miners, we
+require an alternative (albeit less-binding) method to allow light clients to
+''identify'' and ''reject'' invalid filters. The purely p2p solution to this
+problem is to obtain a deterministic hash-chain of ''each'' filter. This hash
+chain or "filter header chain" is similar to the regular Bitcoin headers in
+that it allows a light client to verify the ''authenticity'' of a received
+filter.
+
+The filter header chain for a particular filter type is described by the
+following recurrence:
+<pre>
+filter_header(n: uint) -> [32]byte = 
+   let zero_hash [32]byte = {0*32}
+
+   if n == 0:
+       double-sha-256(genesis_block.prevblock || filter(0))
+
+   match filter(n):
+      case Some:
+          double-sha-256(filter_header(n-1) || double-sha-256(filter(n)))
+      case None:
+          double-sha-256(filter_header(n-1) || double-sha-256(zero_hash))
+
+   where filter(n) is the filter for block height n
+</pre>
+
+The filter header for the genesis block uses the hash stored in the prevblock
+field of the genesis block header itself, as there's no prior filter header
+(by definition).
+
+Due to the nature of filter construction, it's possible to construct a block
+such that an "empty" filter will be produced. This is the case of a coinbase
+transaction that has no data pushes in its public key script. In this case, the
+"hash" of said filter is simply "32 zeroes". 
+
+This filter header chain can be utilized by light clients to gain a greater
+degree of security against bamboozling full-nodes during their initial chain
+sync. In addition to fetching all the bitcoin headers, light clients
+implementing this BIP should also fetch all the ''filter headers'' from ''each'' of
```

Comment:
I think this section may be clearer if you use RFC words (which you appear to already be doing) and be clear that you are. Also, what is the advantage of fetching the filter headerS from each peer? It would be less bandwidth and just as much cross-checking to only fetch the top block's filter header after you've synced the headers chain (which is probably a much better approach in any case as you can first make sure you're on the best chain, and then download a filters chain, protecting you from some types of DoS issues). You can then do a backwards sync of the filters. If there is a mismatch you do a binary search or otherwise to figure out which peer is lying.

Reply:
Modified to make the usage of RFC works explicit. The section below assumes the client has fully synced, as is fetching from tip. As a result, they already have the entire header chain. The naive way would be just to fetch the _entire_ filter from each peer. This algo avoids that and only fetches the filter in the case a bamboozlement attempt by a peer.